service/kernel/oneapi/reducer.h

```diff
@@ -0,0 +1,106 @@
+/* file: reducer.h */
+/*******************************************************************************
+* Copyright 2014-2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __REDUCER_H__
+#define __REDUCER_H__
+
+#include "service/kernel/oneapi/math_service_types.h"
+#include "services/buffer.h"
+#include "service/kernel/oneapi/cl_kernels/sum_reducer.cl"
+#include "oneapi/internal/types_utils.h"
+#include "oneapi/internal/execution_context.h"
+
+namespace daal
+{
+namespace oneapi
+{
+namespace internal
+{
+namespace math
+{
+class SumReducer
+{
+public:
+    struct Result
+    {
+        UniversalBuffer sum;
+        UniversalBuffer sumOfSquares;
+
+        Result(ExecutionContextIface & context, uint32_t nVectors, TypeId type, services::Status * status)
+            : sum(context.allocate(type, nVectors, status)), sumOfSquares(context.allocate(type, nVectors, status))
+        {}
+    };
+
+public:
+    static Result sum(Layout vectorsLayout, const UniversalBuffer & vectors, uint32_t nVectors, uint32_t vectorSize, services::Status * status);
+
+private:
+    SumReducer();
+};
+
+class Reducer
+{
+public:
+    enum class BinaryOp
+    {
+        MIN,
+        MAX,
+        SUM,
+        SUMS_OF_SQUARED
+    };
+
+    struct Result
+    {
+        UniversalBuffer reduceRes;
+
+        Result(ExecutionContextIface & context, uint32_t nVectors, TypeId type, services::Status * status)
+            : reduceRes(context.allocate(type, nVectors, status))
+        {}
+
+        Result(ExecutionContextIface & context, UniversalBuffer & resReduce, uint32_t nVectors, TypeId type, services::Status * status)
+            : reduceRes(resReduce)
+        {}
+    };
+
+public:
+    static Result reduce(const BinaryOp op, Layout vectorsLayout, const UniversalBuffer & vectors, uint32_t nVectors, uint32_t vectorSize,
```

Comment:
Why during runtime parameter? This will be build OpenCL program with one or another functor. During startup, all functions inside are compiled. I think when we go to the sycl kernels it will be prettier.

Reply:
Now, `op` is used for dispatching in runtime with the help of `if`/`switch`. I'm trying to understand relevance of runtime dispatching here. I guess, we always know the type of reduction operation in compile time, so the switches wrt `op` and `vectorsLayout` should be replaced by templates in new API.

Reply:
Just a comment to address in the future.

Reply:
I think it would look cool. But I don’t think that this will play a key role for the performance.